File: src/main/java/ru/piko/pikopluginlib/Listeners/PluginListener.java
package ru.piko.pikopluginlib.Listeners;

import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.server.PluginDisableEvent;
import org.bukkit.event.server.PluginEnableEvent;
import org.bukkit.plugin.java.JavaPlugin;
import ru.piko.pikopluginlib.Main;
import ru.piko.pikopluginlib.PikoPlugin;
import ru.piko.pikopluginlib.PikoPluginData;

public class PluginListener implements Listener {

    @EventHandler
    public void onEnablePlugin(PluginEnableEvent e) {
        if (e.getPlugin() instanceof JavaPlugin javaPlugin && javaPlugin instanceof PikoPlugin plugin) {
            if (!Main.getPlugin().hasPikoPlugin(plugin.getPluginId())) {
                PikoPluginData data = Main.getPlugin().getPikoPluginData(plugin.getPluginId());
                if (data != null) { data.activate(plugin, false); }
            } else {
                Main.getPlugin().addPikoPlugin(plugin.getPluginId(), plugin);
            }
        }
    }

    @EventHandler
    public void onDisablePlugin(PluginDisableEvent e) {
        if (e.getPlugin() instanceof JavaPlugin javaPlugin && javaPlugin instanceof PikoPlugin plugin) {
            PikoPluginData data = Main.getPlugin().getPikoPluginData(plugin.getPluginId());
            if (data == null) {
                Main.getPlugin().getLogger().warning("How did this even happen???!!!");
                Main.getPlugin().addDisablePikoPlugin(plugin.getPluginId());
                data = Main.getPlugin().getPikoPluginData(plugin.getPluginId());
            } else {
                data.disable();
            }
            String name = data.getNamePlugin();
            if (name == null) name = "???";
            Main.getPlugin().getLogger().info("The " + name + "plugin is disabled. The status in the PikoPlugins system is set to disable.");
        }
    }

}

File: src/main/java/ru/piko/pikopluginlib/Listeners/PluginListener.kt
package ru.piko.pikopluginlib.Listeners

import org.bukkit.Bukkit
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.server.PluginDisableEvent
import org.bukkit.plugin.java.JavaPlugin
import ru.piko.pikopluginlib.Functions.FunctionTimer
import ru.piko.pikopluginlib.Api.PikoPlugin
import ru.piko.pikopluginlib.Utils.InternalObject.main

class PluginListener : Listener {
	
	@EventHandler
	fun onDisablePlugin(e: PluginDisableEvent) {
		val plugin = e.plugin
		if (plugin is JavaPlugin && plugin is PikoPlugin) {
			
			// Проверка, что плагин не в процессе первичной загрузки
			if (plugin.pluginLoadingInProgress) return
			
			var data = main.api.plugins.get(plugin.id)
			if (data == null) {
				main.logger.warning("How could this even happen???!!! A plugin that is not in the system is disabled. How did it bypass the activation? is the PikoPluginLib duplicate shaded?")
				main.api.plugins.addDisable(plugin.id)
				data = main.api.plugins.get(plugin.id)
			} else {
				if (!data.status.isDisable) {
					if (data.plugin != plugin) {
						throw IllegalStateException("Identical IDs for different instances! One: ${data.file.toString()}; Two: ${plugin.pluginFile}")
					}
				}
				if (data.status.isBlocked) {
					data.disable()
					main.logger.info("The ${data.namePlugin} plugin is disabled. The status in the PikoPlugins system is set to 'BLOCKED_DISABLE'.")
					
					try {
						error("call catch")
					} catch (_: Exception) {
						throw IllegalStateException("Disabling a blocked plugin. Blocked plugins cannot be disabled!")
					} finally {
						Bukkit.shutdown()
					}
				}
			}
			val name = data?.namePlugin ?: "???"
			main.logger.info("The $name plugin is disabled. The status in the PikoPlugins system is set to disable.")
		}
	}
	
}